### PR TITLE
Add gpu dev notes

### LINKED ISSUE #1356: How does the CUDA Jit function?
I cannot find any documentation on its operation, or how to make it show me exactly what it is doing. I have a very simple PETSc project, which I am uploading. Wait, Github dos not support attaching C files or makefiles. Ridiculous. Anyway, here is the tarball.

[ex11.tgz](https://github.com/CEED/libCEED/files/12785500/ex11.tgz)

The problem is that when we select a CUDA backend

```
make run_sw EXTRA_OPTIONS="-dm_ceed /gpu/cuda/ref -cdm_dm_ceed /gpu/cuda/ref -dm_vec_type cuda -cdm_vec_type cuda"
```

we get an error from the CUDA compiler

```
/home/arcowie/petsc/arch-linux-rdycore-debug/externalpackages/git.libceed/backends/cuda/ceed-cuda-compile.cpp:93 in CeedCompile_Cuda(): NVRTC_ERROR_COMPILATION
default_program(371): error: identifier "PetscInt" is undefined
    PetscInt    dof;
    ^
```

which clearly happens because the compiler cannot see the includes. How do I tell it where the includes are?

### PR DESCRIPTION
**Pull request by jeremylt**
Closes #1356 

This MR adds additional documentation to clarify a couple of common points about GPU development.

### CODE REVIEW

File: doc/sphinx/source/gpu.md [hunk 1]
```diff
@@ -0,0 +1,22 @@
+# GPU Development
+
+Runtime selection of libCEED backends allows users to use CPU backends for easier debugging.
+Code that produces correct results with CPU backends will produce correct results on GPU backends, provided that JiT and memory access assumptions of the libCEED API are respected.
+
+## JiT Compilation
+
+The filepath to the user source code is passed in {c:func}`CeedQFunctionCreateInterior` as the `source` argument.
+This filepath should typically be an absolute path to ensure the JiT compilation can locate the source file.
```
Comment (jrwrigh) [inline doc/sphinx/source/gpu.md:9]: Maybe also mention that the `CEED_QFUNCTION` automatically creates this path in a string variable?

File: doc/sphinx/source/gpu.md [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+The entire contents of this file and all locally included files (`#include "foo.h"`) are used during JiT compilation for GPU backends.
+Installed headers (`#include <bar.h>`) are omitted in the source code passed to JiT, but the compilation environment may supply common headers such as `<math.h>`.
+These source file must only contain syntax constructs supported by C99 and all targeted backends (i.e. CUDA for `/gpu/cuda`, OpenCL/SYCL for `/gpu/sycl`, etc.).
+
+All source files must be at the provided filepath at runtime for JiT to function.
+
+## Memory Access
+
+GPU backends require stricter adherence to memory access assumptions, but CPU backends may occasionally report correct results despite violations of memory access assumptions.
```
Comment (jrwrigh) [inline doc/sphinx/source/gpu.md:20]: Maybe explicitly mention QFunction contexts as an example? I don't really know of another example from user-facing code tbh.

**Review by jrwrigh — APPROVED** [review]

File: doc/sphinx/source/gpu.md (hunk 2 above)
Comment (jrwrigh) [reply doc/sphinx/source/gpu.md:20]: Ah, yeah. Forgot about that one, just didn't trip over it nearly as much as the context.

File: doc/sphinx/source/gpu.md (hunk 1 above)
Comment (jeremylt) [reply doc/sphinx/source/gpu.md:9]: added

File: doc/sphinx/source/gpu.md (hunk 2 above)
Comment (jeremylt) [reply doc/sphinx/source/gpu.md:21]: I added some expanded explanatory text

**Pull request merged**